drivers/adc/adc_nrfx_saadc.c

```diff
@@ -368,38 +412,44 @@ static int adc_nrfx_read_async(const struct device *dev,
 
 static void saadc_irq_handler(const struct device *dev)
 {
-	if (nrf_saadc_event_check(NRF_SAADC, NRF_SAADC_EVENT_END)) {
-		nrf_saadc_event_clear(NRF_SAADC, NRF_SAADC_EVENT_END);
+	nrfx_saadc_evt_type_t event;
+	nrfx_err_t err;
 
-		nrf_saadc_task_trigger(NRF_SAADC, NRF_SAADC_TASK_STOP);
-		nrf_saadc_disable(NRF_SAADC);
+	event = m_data.event->type;
 
+	if (event == NRFX_SAADC_EVT_DONE) {
 		adc_context_on_sampling_done(&m_data.ctx, dev);
-	} else if (nrf_saadc_event_check(NRF_SAADC,
-					 NRF_SAADC_EVENT_CALIBRATEDONE)) {
-		nrf_saadc_event_clear(NRF_SAADC, NRF_SAADC_EVENT_CALIBRATEDONE);
-
-		/*
-		 * The workaround for Nordic nRF52832 anomalies 86 and
-		 * 178 is an explicit STOP after CALIBRATEOFFSET
-		 * before issuing START.
-		 */
-		nrf_saadc_task_trigger(NRF_SAADC, NRF_SAADC_TASK_STOP);
-		nrf_saadc_task_trigger(NRF_SAADC, NRF_SAADC_TASK_START);
-		nrf_saadc_task_trigger(NRF_SAADC, NRF_SAADC_TASK_SAMPLE);
+	} else if (event == NRFX_SAADC_EVT_CALIBRATEDONE) {
```

Comment:
this event will never arrive as `nrfx_saadc_offset_calibrate();` is used with NULL parameter, therefore it executes in blocking manner (i.e. no interrupt is used if handler for calibration is not provided)